### PR TITLE
Add unit tests for db/repos/blocks - Closes #1565

### LINKED ISSUE #1565: Add unit test coverage for db/repos/blocks
Parent: #1484 

### Expected behavior
We need 100% coverage for db/repos/blocks

### Actual behavior
Currently there is no coverage for this repo.

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?

No unit test coverage for db/repos/blocks

### How did I fix it?

Added test cases thoroughly for every method in db/repos/blocks. And got 100% coverage for that file.

### How to test it?
```
npm mocha test/unit/db/repos/blocks.js
```

### Review checklist

* The PR solves #1565
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

**Comment by ManuGowda** [comment]
@nazarhussain can you please pull the laster from `1.0.0` and run `npm run lint` this test `test/unit/db/repos/blocks.js` has synchronus test errors.

**Comment by nazarhussain** [comment]
@ManuGowda Thanks doing it.

File: test/unit/db/repos/blocks.js [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+
+	after(done => {
+		dbSandbox.destroy();
+		done();
+	});
+
+	beforeEach(done => {
+		seeder
+			.seed(db)
+			.then(() => done(null))
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: why not just `done()`?

File: test/unit/db/repos/blocks.js [hunk 2]
```diff
@@ -0,0 +1,885 @@
...
+			.catch(done);
+	});
+
+	it('should initialize db.blocks repo', done => {
+		expect(db.blocks).to.be.not.null;
+		done();
+	});
+
+	describe('AccountsRepository', () => {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: `BlocksRepository` ofc :)

File: test/unit/db/repos/blocks.js [hunk 3]
```diff
@@ -0,0 +1,885 @@
...
+					'blockSignature'
+				);
+
+				const payloadHash = genesisBlock[0].payloadHash.toString('hex');
+
+				return expect(payloadHash).to.be.eql(__testContext.config.nethash);
+			});
+		});
+
+		describe('getGenesisBlockId()', () => {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: Consistency - `getGenesisBlockId` or `getGenesisBlockId(id)` (I like first more)

File: test/unit/db/repos/blocks.js [hunk 4]
```diff
@@ -0,0 +1,885 @@
...
+
+				expect(db.none.firstCall.args[0]).to.eql(sql.deleteBlock);
+				expect(db.none.firstCall.args[1]).to.eql(['1111111']);
+				return expect(db.none).to.be.calledOnce;
+			});
+
+			it('should delete a block for given id if it exists', function*() {
+				const genesisBlock = blocksFixtures.GenesisBlock();
+				const before = yield db.one(
+					'SELECT count(*) from blocks WHERE id = $1',
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: block id is unique - `count` here doesn't make much sense

File: test/unit/db/repos/blocks.js [hunk 5]
```diff
@@ -0,0 +1,885 @@
...
+				);
+			});
+		});
+
+		describe('aggregateBlocksReward()', () => {
+			it('should use the correct SQL file with three parameters', function*() {
+				sinonSandbox.spy(db, 'query');
+				const params = {
+					generatorPublicKey: 'afafafafaf',
+					start: (+new Date() / 1000).toFixed(0),
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: number of digits is optional, and default `0`, so just `.toFixed()` like everywhere in our system

File: test/unit/db/repos/blocks.js [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+			});
+
+			it('should not throw an error if a block with given id does not exists', () => {
+				return expect(db.blocks.deleteBlock('111111')).to.be.eventually.equal(
+					null
+				);
+			});
+		});
+
+		describe('aggregateBlocksReward()', () => {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: please add test cases:
- valid, but not existing (in database) public key
- existing public key, but account is not a delegate

File: test/unit/db/repos/blocks.js [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+					'b_previousBlock',
+					'b_reward',
+					'b_timestamp',
+					'b_totalAmount',
+					'b_totalFee',
+					'b_version',
+				]);
+			});
+
+			it('should paginate the blocks properly', function*() {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: fragile, sort by unique field should be used as we cannot expect postgres to always sort same way
but maybe for that small data set can be ok...

File: test/unit/db/repos/blocks.js [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+				});
+
+				expect(result.length).to.be.above(0);
+
+				return result.forEach(block => {
+					expect(block.b_id).to.be.eql(genesisBlock.id);
+				});
+			});
+
+			it('should return valid blocks with composite conditions', function*() {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: valid blocks =/= empty array

File: test/unit/db/repos/blocks.js [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+				const genesisBlock = blocksFixtures.GenesisBlock();
+				const result = yield db.blocks.list({
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: please add comment if where conditions are joined by `AND` or `OR`

File: test/unit/db/repos/blocks.js [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+				const commonBlock = yield db.blocks.getCommonBlock(params);
+
+				expect(commonBlock).to.be.not.empty;
+				expect(commonBlock).to.be.an('array');
+				expect(commonBlock).to.be.lengthOf(1);
+				expect(commonBlock[0]).to.have.all.keys('count');
+				return expect(commonBlock[0].count).to.be.eql(1);
+			});
+
+			it('should return the count of blocks matching "previousBlock" condition', function*() {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: only `previousBlock` or also `id` and `height`?

File: test/unit/db/repos/blocks.js [hunk 11]
```diff
@@ -0,0 +1,885 @@
...
+				expect(db.query.firstCall.args[0]).to.eql(sql.getHeightByLastId);
+				expect(db.query.firstCall.args[1]).to.eql(['11111']);
+				return expect(db.query).to.be.calledOnce;
+			});
+
+			it('should be fulfilled with empty response if required param "lastId" is missing', () => {
+				return expect(db.blocks.getHeightByLastId()).to.be.eventually.eql([]);
+			});
+
+			it('should correct height of the given block id', function*() {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: `should return correct...`

File: test/unit/db/repos/blocks.js [hunk 12]
```diff
@@ -0,0 +1,885 @@
...
+				const block = blocksFixtures.GenesisBlock();
+				const data = yield db.blocks.loadBlocksData({
+					lastId: block.id,
+					height: 1,
+					limit: 10,
+				});
+
+				expect(data).to.be.not.empty;
+				expect(data).to.be.an('array');
+				// There are 5 total blocks in the seed
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: so why we expect result of `4`?

File: test/unit/db/repos/blocks.js [hunk 13]
```diff
@@ -0,0 +1,885 @@
...
+
+			it('should use condition limit condition as height if both "id" and "lastId" is not provided', function*() {
+				const data = yield db.blocks.loadBlocksData({
+					height: 1,
+					limit: 3,
+				});
+
+				expect(data).to.be.not.empty;
+				expect(data).to.be.an('array');
+				// There are 5 total blocks in the seed
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: same here, please improve that comment - query: 3, total: 5, expect: 2 - it need explanation ;)

File: test/unit/db/repos/blocks.js [hunk 14]
```diff
@@ -0,0 +1,885 @@
...
+				const genesisBlock = blocksFixtures.GenesisBlock();
+				const data = yield db.blocks.loadLastNBlockIds(5);
+
+				expect(data).to.be.not.empty;
+				expect(data).to.be.an('array');
+				expect(data).to.have.lengthOf(5); // As look for 5 blocks
+				expect(data[0]).to.have.all.keys('id');
+
+				// Due to descending order on height genesis block will appear in last
+				return expect(data[data.length - 1].id).to.be.eql(genesisBlock.id);
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: can we test all data set here?

File: test/unit/db/repos/blocks.js [hunk 15]
```diff
@@ -0,0 +1,885 @@
...
+				sinonSandbox.spy(db, 'none');
+
+				yield db.blocks.deleteAfterBlock('1111');
+
+				expect(db.none.firstCall.args[0]).to.eql(sql.deleteAfterBlock);
+				expect(db.none.firstCall.args[1]).to.eql(['1111']);
+				return expect(db.none).to.be.calledOnce;
+			});
+
+			it('should delete all blocks having height above given block id', function*() {
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: `all blocks with height`

File: test/unit/db/repos/blocks.js [hunk 16]
```diff
@@ -0,0 +1,885 @@
...
+				);
+				const data = yield db.blocks.deleteAfterBlock(genesisBlock.id);
+				const after = yield db.query(
+					'SELECT height FROM blocks WHERE height >= $1',
+					[genesisBlock.height]
+				);
+
+				expect(data).to.be.null;
+				expect(before).to.have.length(5);
+				return expect(after).to.have.length(0);
```
Comment (4miners) [inline test/unit/db/repos/blocks.js]: Something is wrong, it should not delete genesis block, we should expect `1` here

File: test/unit/db/repos/blocks.js (hunk 3 above)
Comment (nazarhussain) [reply test/unit/db/repos/blocks.js]: I used `getGenesisBlockId()` just to show in tests results that its a function that is tested. To differentiate from testing some data attribute.

File: test/unit/db/repos/blocks.js (hunk 8 above)
Comment (nazarhussain) [reply test/unit/db/repos/blocks.js]: @4miners Can't understand this comment

File: test/unit/db/repos/blocks.js (hunk 14 above)
Comment (nazarhussain) [reply test/unit/db/repos/blocks.js]: We don't keep the track of fixtures we load during seed. We need to refactor that seeding logic to test all blocks.

File: test/unit/db/repos/blocks.js (hunk 16 above)
Comment (nazarhussain) [reply test/unit/db/repos/blocks.js]: Its doing as expected. I updated the test description accordingly.

```
WHERE height >= (SELECT height FROM blocks WHERE id = $1)
```

This is the condition, so genesis block will also be deleted due to `>=` condition.

Comment (4miners) [reply test/unit/db/repos/blocks.js]: Yep, looks like function name is misleading.

File: test/unit/db/repos/blocks.js (hunk 6 above)
Comment (nazarhussain) [reply test/unit/db/repos/blocks.js]: Done

**Pull request merged**